src/linker/pipeline.py

```diff
@@ -16,81 +15,40 @@ class Pipeline:
     def __init__(self, config: Config):
         self.config = config
         self.steps = self._get_steps()
-        self.implementation_metadata = self._load_implementation_metadata()
-        self.runner = None  # Assigned later from set_runner method
-
-    def set_runner(self, runner: Callable) -> None:
-        self.runner = runner
+        self.implementations = self._get_implementations()
 
-    def run(self, results_dir: Path):
-        if not self.runner:
-            raise RuntimeError("Runner has not been set.")
-        for step_name in self.steps:
-            self.runner(
+    def run(self, runner: Callable, results_dir: Path) -> None:
+        for implementation in self.implementations:
+            implementation.run(
+                runner=runner,
                 container_engine=self.config.container_engine,
                 input_data=self.config.input_data,
                 results_dir=results_dir,
-                step_name=step_name,
-                implementation_dir=self._get_implementation_directory(step_name),
-                container_full_stem=self._get_container_full_stem(step_name),
             )
 
+    def _validate(self, schema: "PipelineSchema") -> None:
+        for idx, implementation in enumerate(self.implementations):
+            # Check that all steps are accounted for and in the correct order
+            if implementation.step_name != schema.steps[idx].name:
+                return False
+            # Check that container exists
```

Comment:
This could be checked on instantiating the Implementation, though I don't feel strongly.

Reply:
+1. Add a validate method that is called during Implementation's init.